### PR TITLE
[draft] Add some missing write barriers when dealing with global caches

### PR DESCRIPTION
**Pull request by jaykrell**
accessed from multiple threads without locks.

Note that read barriers might also be needed.
But maybe not, due to data dependencies.
Testing on weak ARM hardware needed.
And mono doesn't have read barriers, just full barriers by another name.
Does ARM?

### CODE REVIEW

**Comment by jaykrell** [comment]
@monojenkins build failed

File: mono/utils/mono-dl.c [hunk 1]
```diff
@@ -322,8 +322,11 @@ mono_dl_symbol (MonoDl *module, const char *name, void **symbol)
 	}
 
 	if (sym) {
-		if (symbol)
+		if (symbol) {
+			// Caller is often but not always caching in a global.
+			mono_memory_barrier ();
```
Comment (BrzVlad) [inline mono/utils/mono-dl.c]: this barrier doesn't seem to belong

**Comment by BrzVlad** [comment]
This PR is too big. Some changes also seem unrelated or they don't follow any clear pattern, making reviewing awkward. Maybe we should start with only the static locals and take it from there (MONO_STATIC_POINTER_INIT). Note that this PR adds excessive memory barriers, mono_memory_barrier is always a full barrier, even on x86/amd64 where we don't need it. We should only make sure we have release semantics when storing to the static field (either use STORE_RELEASE_FENCE or mono_atomic_store_release, which also has equivalent asm instruction on arm64 and which is NOP on amd64). We should really stop using those mono_memory_barrier methods in the runtime. Seems fair that we won't be needing any read barriers there since we have data dependency when loading the contents of the ptr. Hopefully ..

File: mono/utils/mono-dl.c (hunk 1 above)
Comment (jaykrell) [reply mono/utils/mono-dl.c]: I can review the callers instead.

**Comment by jaykrell** [comment]
Agreed `mono_memory_barrier` is overkill, but it seemed to be prevailing practice.
The fences appear inadequate, in that they lack compiler fencing, at least on S390x.
I do now see:
```
TODO: replace all explicit uses of memory barriers with macros from this section
```

**Comment by jaykrell** [comment]
Can you clarify which to do first vs. later?

It is ambiguous.

I guess I suggest, first:

introduce MONO_STATIC_POINTER_INIT
with either zero, some, or many uses.

Anything that is not MONO_STATIC_POINTER_INIT, even if similar,
leave for separate.

I think we should also introduce, separate PR:
 `mono_compiler_barrier ()`;

which
is like Linux's asm ("") or such, with gcc,
and there is an intrinsic for msvc.

It generates zero code, but inhibits compiler moving
code across. A full barrier, in this cheap way.
It inhibits no processor reordering.

It is generally useless on its own.
It should be part of all release/store fences, even the empty ones.

I will leave this PR alone or closed, and open a new one based on it.

**Comment by BrzVlad** [comment]
Let's handle for starters all the static local initialization cases where barriers are currently missing using that MONO_STATIC_POINTER_INIT.

https://github.com/mono/mono/pull/16461 will add a compiler barrier. We can reuse that later as necessary. I suggest replacing empty fences from `mono-memory-model.h` with this compiler barrier. In the cases where the fences are defined there shouldn't be anything left to do, the compiler barrier should be implicit.

**Comment by jaykrell** [comment]
compiler barrier: Yes, that is exactly what I want, good.
And replacing the empties, agreed.
Not clear if the "processor barriers" imply the "compiler" ones or not.

MONO_STATIC_POINTER_INIT: I guess the implication is that
some of the cases already had locals/barriers so were ok,
making them consistent can be separate, but I still think
should be done at some point.

File: mono/metadata/class-init.c [hunk 2]
```diff
@@ -4128,7 +4122,8 @@ mono_class_get_object_finalize_slot ()
 MonoMethod *
 mono_class_get_default_finalize_method ()
 {
-	return default_finalize;
+	int const i = finalize_slot;
+	return (i < 0) ? NULL : mono_defaults.object_class->vtable [i];
 }
```
Comment (naricc) [inline mono/metadata/class-init.c]: Is this related to the write barrier changes?

Comment (jaykrell) [reply mono/metadata/class-init.c]: This PR is a smaller form, yes.
And this code initializes two globals, with one guard.
My change is to initialize one global, with one guard.

**Comment by naricc** [comment]
So, is the issue here that ARM has a weaker memory model and we might get word shear or some similar crazy thing while doing lazy initialization the way we have been?

**Comment by jaykrell** [comment]
Yes.

**Comment by lambdageek** [comment]
Okay so the plan is:

1. a new PR with just the definition (and maybe 1 use) of `MONO_STATIC_POINTER_INIT`/`_END` with a `STORE_RELEASE_FENCE` https://github.com/mono/mono/pull/17849#issuecomment-556055751
2. several small follow-up PRs to change `static` locals initialization to use the macros where they currently don't use any barrier at all.
3. At a later point, after #16461 lands, the no-op fence macros are updated to add a compiler barrier (If I'm reading https://github.com/mono/mono/pull/17849#issuecomment-556273801 correctly - hardware fences already imply compiler fences, so it's only the architectures with a strong memory ordering (where the fences are empty) that will need the compiler barriers)
4. The cases where a static is initialized and a barrier is already included, we don't change it.

@jaykrell I'm going to close this PR now.

**Comment by jaykrell** [comment]
I think step 4. should be, in the cases that already ok, but easily converted to `MONO_STATIC_POINTER_INIT`, convert. As this PR did.

There aren't a ton, and the consistency and removal of repetition is worthwhile.
